app/src/main/java/ch/epfl/culturequest/ui/scan/CameraSetup.java

```diff
@@ -0,0 +1,341 @@
+     * @param textureView the texture view from the activity
+     */
+    public CameraSetup(CameraManager cameraManager, TextureView textureView) {
+        this.cameraManager = cameraManager;
+        this.textureView = textureView;
+
+        // We create a background thread to handle the camera
+        backgroundThread = new HandlerThread("Camera feedback");
+        openThread();
+    }
+
+    /**
+     * Opens the Camera and starts the preview
+     */
+    public void openCamera() {
+        try {
+            String cameraId = "0";
+
+            //We open the camera
+            cameraManager.openCamera(cameraId, cameraState, backgroundHandler);
+        } catch (SecurityException | CameraAccessException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Requests the picture
+     */
+    public CompletableFuture<Boolean> takePicture() {
+        if (cameraDevice == null || photoRequest == null || imageReader == null) {
+            return CompletableFuture.completedFuture(false);
+        }
+
+        CompletableFuture<Boolean> future = new CompletableFuture<>();
+
+        try {
+            //We start the capture session
+            CameraCaptureSession.CaptureCallback captureCallback = new CameraCaptureSession.CaptureCallback() {
+                @Override
+                public void onCaptureCompleted(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull TotalCaptureResult result) {
+                    super.onCaptureCompleted(session, request, result);
+                    future.complete(true);
+                }
+            };
+            cameraCaptureSession.capture(photoRequest, captureCallback, null);
+        } catch (CameraAccessException e) {
+            e.printStackTrace();
+            return CompletableFuture.completedFuture(false);
+        }
+        return future;
+    }
+
+    /**
+     * Gets the latest image taken
+     */
+    public CompletableFuture<Bitmap> getLatestImage() {
+        //We wait for the image to be read
+        while (latestImage == null) {
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        //We convert the image to a bitmap
+        Image.Plane[] planes = latestImage.getPlanes();
+        byte[] data = new byte[planes[0].getBuffer().remaining()];
+        planes[0].getBuffer().get(data);
+        Bitmap bitmap = BitmapFactory.decodeByteArray(data, 0, data.length, null);
+
+        //We rotate the image
+        Matrix matrix = new Matrix();
+        matrix.postRotate(90);
+        bitmap = Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
+
+        //We close the image to free up memory
+        latestImage.close();
+        latestImage = null;
+
+        return CompletableFuture.completedFuture(bitmap);
+    }
+
+
```

Comment:
Maybe creating a new camera package with this CameraSetup class in it and creating a new file for another class that extends CameraDevice.StateCallback could be a good idea. Because this CameraDevice.StateCallback class anonymous declaration is so big that it may deserve its own class.

Reply:
Why not, but an entire package for only two classes seems too much to me, considering the fact that it is only used for the scan UI.

Reply:
I just saw that moving the StateCallback means moving most of the methods in the second class, or adding a high number of parameters (there is a strong dependence between the data used in the StateCallback and the CameraSetup). Maybe it is not worth it.

Reply:
Ok no worries, if these two classes are really dependent then let it like that :)